File: rust/src/bin/day01.rs
use std::{collections::BTreeMap, error::Error, io::BufRead};

use rust::read;

fn read_input() -> Result<(Vec<i32>, Vec<i32>), Box<dyn Error>> {
    let buf = read(1, false)?;
    let res = buf
        .lines()
        .map(|line| {
            let items: Vec<i32> = line
                .expect("line must be read")
                .split_whitespace()
                .map(|s| s.parse::<i32>().expect("parse success"))
                .take(2)
                .collect();
            (items[0], items[1])
        })
        .unzip();

    Ok(res)
}

fn part1() -> Result<(), Box<dyn Error>> {
    let (mut arr1, mut arr2) = read_input()?;
    arr1.sort();
    arr2.sort();
    let ans = arr1
        .into_iter()
        .zip(arr2)
        .map(|(v1, v2)| (v1 - v2).abs() as i64)
        .sum::<i64>();

    println!("{}", ans);
    Ok(())
}

fn part2() -> Result<(), Box<dyn Error>> {
    let (arr1, arr2) = read_input()?;
    let map = arr2.into_iter().fold(BTreeMap::new(), |mut map, e| {
        *map.entry(e).or_insert(0) += 1;
        map
    });
    let ans = arr1
        .into_iter()
        .map(|v| (v as i64) * map.get(&v).unwrap_or(&0))
        .sum::<i64>();

    println!("{}", ans);
    Ok(())
}

fn main() -> Result<(), Box<dyn Error>> {
    part1()?;
    part2()?;
    Ok(())
}

File: python/day01.py
from collections import defaultdict

from utils import read


def part1() -> None:
    arr1, arr2 = [], []
    with read(day=1, sample=False) as f:
        for line in f.readlines():
            p1, p2 = map(int, line.split())
            arr1.append(p1)
            arr2.append(p2)
        arr1 = sorted(arr1)
        arr2 = sorted(arr2)

    s = sum(abs(p1 - p2) for p1, p2 in zip(arr1, arr2, strict=False))
    print(s)


def part2() -> None:
    arr1 = []
    dict2: dict[int, int] = defaultdict(lambda: 0)
    with read(day=1, sample=False) as f:
        for line in f.readlines():
            p1, p2 = map(int, line.split())
            arr1.append(p1)
            dict2[p2] += 1
        arr1 = sorted(arr1)

    s = sum(p * dict2[p] for p in arr1)
    print(s)


if __name__ == "__main__":
    part1()
    part2()
